### PR TITLE
Fix logrus logger hook data race condition

### PR DESCRIPTION
**Pull request by bjsignalfx**
Data race issue https://github.com/signalfx/splunk-otel-collector/pull/166#issuecomment-801643031 fixed by updating logrus to version v1.8.1. Plus refactoring.

### CODE REVIEW

**Comment by rmfitzpatrick** [comment]
I don't believe this resolves the noticed issue: https://github.com/signalfx/splunk-otel-collector/pull/166#issuecomment-801643031

If you merge that branch and run `go test -failfast -race -run TestMultipleStartAndShutdownInvocations -count 10000` it occurs with and without these changes.

**Comment by bjsignalfx** [comment]
@rmfitzpatrick you are correct. Nonetheless it looks like this problem is fixed in logrus v1.8.1 (see [here](https://github.com/sirupsen/logrus/blob/v1.8.1/entry.go#L263)). I cleanup a few things and added tests. Can you please review those changes?

**Comment by bjsignalfx** [comment]
@rmfitzpatrick I forgot to mention that we are using v1.8.0.

**Comment by rmfitzpatrick** [comment]
@bjsignalfx would just bumping the logrus version be able to resolve on its own?

**Comment by bjsignalfx** [comment]
@rmfitzpatrick yes that resolves the data race condition but now the error is:
`goroutine 23150 [semacquire, 1 minutes]:
github.com/shirou/gopsutil/cpu.allCPUTimes(0x61eb6e5, 0x4010ceb, 0x70d6250, 0x0, 0x0)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin_cgo.go:87 +0x55
github.com/shirou/gopsutil/cpu.TimesWithContext(0x8427050, 0xc0001a4008, 0xc00b2df800, 0xc00007a900, 0x4011458, 0xc0024c6f00, 0xc00b2df860, 0xc006ca0cd0)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin.go:44 +0xcb
github.com/shirou/gopsutil/cpu.Times(0x9813b00, 0x18, 0x0, 0xffffffffffffffff, 0xc006ca0d70, 0x403e8c5)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin.go:36 +0x65
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generateDatapoints(0xc016436480, 0xc006ca0f20, 0xc006512ee0, 0xc005856000)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:108 +0x65
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:243 +0x76
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc007391a60, 0xc0119281e0, 0x8427018, 0xc016436500)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23266 [semacquire, 1 minutes]:
github.com/signalfx/signalfx-agent/pkg/utils.MergeStringMaps(0xc00c719c58, 0x2, 0x2, 0x3)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/maps.go:27 +0xf4
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generatePerCoreDatapoints(0xc00445c140, 0xc011041e80, 0xa, 0x10)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:93 +0x2ee
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:247 +0x12d
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc006d5e180, 0xc00b1e2c80, 0x8427018, 0xc00f1c5680)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23200 [semacquire, 1 minutes]:
github.com/signalfx/golib/v3/datapoint.New(...)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/golib/v3@v3.3.16/datapoint/datapoint.go:179
github.com/signalfx/golib/v3/sfxclient.CumulativeF(...)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/golib/v3@v3.3.16/sfxclient/datapointcreation.go:27
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.makeSecondaryDatapoints(0xc00c365ca8, 0xc01a810af0, 0x9, 0xc01a67bc80)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:162 +0x7a5
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generatePerCoreDatapoints(0xc008f7a400, 0xc01391bf80, 0xa, 0x10)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:72 +0x60a
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:247 +0x12d
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc0047be700, 0xc011c51270, 0x8427018, 0xc008f7a580)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23151 [runnable]:
github.com/shirou/gopsutil/cpu._Cfunc_vm_deallocate(0x103, 0x606af000, 0x8, 0x0)
        _cgo_gotypes.go:150 +0x6b
github.com/shirou/gopsutil/cpu.perCPUTimes(0xc0151e8000, 0x8, 0x8, 0x0, 0x0)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin_cgo.go:83 +0x825
github.com/shirou/gopsutil/cpu.TimesWithContext(0x8427050, 0xc0001a4008, 0xc00c361b01, 0xc004b230e0, 0xd, 0x0, 0x0, 0x0)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin.go:41 +0x45
github.com/shirou/gopsutil/cpu.Times(0x9813b01, 0x4045bbc, 0xc015bc8480, 0x300000002, 0xc015bc8480, 0xc00c361b20)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin.go:36 +0x65
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generatePerCoreDatapoints(0xc0117e4040, 0xc01366b600, 0xa, 0x10)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:56 +0x65
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:247 +0x12d
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc004beb240, 0xc0163611d0, 0x8427018, 0xc0117e40c0)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23282 [semacquire, 1 minutes]:
encoding/binary.Read(0x83d9ac0, 0xc018865d10, 0x843a148, 0xa2adf40, 0x7239cc0, 0xc00b01ada0, 0x1, 0x2)
        /usr/local/opt/go/libexec/src/encoding/binary/binary.go:255 +0x250
github.com/shirou/gopsutil/cpu.perCPUTimes(0x0, 0x0, 0x0, 0x0, 0x0)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin_cgo.go:67 +0x497
github.com/shirou/gopsutil/cpu.TimesWithContext(0x8427050, 0xc0001a4008, 0xc018d9bb01, 0xc00c485190, 0xd, 0x0, 0x0, 0x0)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin.go:41 +0x45
github.com/shirou/gopsutil/cpu.Times(0x9813b01, 0x4045bbc, 0xc015bc8f00, 0x300000002, 0xc015bc8f00, 0xc018d9bb20)
        /Users/bjonathan/go/pkg/mod/github.com/shirou/gopsutil@v3.21.2+incompatible/cpu/cpu_darwin.go:36 +0x65
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generatePerCoreDatapoints(0xc014c21b80, 0xc0185cb300, 0xa, 0x10)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:56 +0x65
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:247 +0x12d
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc004924720, 0xc00e4abb30, 0x8427018, 0xc014c21c00)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23251 [semacquire, 1 minutes]:
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generateDatapoints(0xc005721900, 0xc0095d1f20, 0xc00478d5a0, 0xc004575800)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:137 +0xedf
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:243 +0x76
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc003bcd4c0, 0xc007e2a230, 0x8427018, 0xc00e775340)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23271 [semacquire, 1 minutes]:
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure.func1()
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:248 +0x1f2
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc006d5f2a0, 0xc001f37a90, 0x8427018, 0xc00d4951c0)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:56 +0x9c
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23158 [select]:
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc004bea140, 0xc01b60eff0, 0x8427018, 0xc002746c80)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:52 +0x174
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23159 [select]:
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc004beb7c0, 0xc015e29c70, 0x8427018, 0xc004b5da00)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:52 +0x174
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79

goroutine 23243 [runnable]:
time.startTimer(0xc020d4a918)
        /usr/local/opt/go/libexec/src/runtime/time.go:212 +0x47
time.NewTicker(0x3b9aca00, 0x4011458)
        /usr/local/opt/go/libexec/src/time/tick.go:39 +0x15a
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval(0x8427018, 0xc01a93e980, 0xc0047be220, 0x3b9aca00)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:44 +0x39
github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).Configure(0xc01a93e900, 0xc00354fe00, 0x0, 0x0)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:242 +0x3f5
reflect.Value.call(0x759c360, 0xc01a93e900, 0x213, 0x78363ba, 0x4, 0xc002e23a10, 0x1, 0x1, 0x410f991, 0x741ca77, ...)
        /usr/local/opt/go/libexec/src/reflect/value.go:476 +0x99c
reflect.Value.Call(0x759c360, 0xc01a93e900, 0x213, 0xc002e23a10, 0x1, 0x1, 0xc01a93e900, 0x213, 0xc006ae52d0)
        /usr/local/opt/go/libexec/src/reflect/value.go:337 +0xd9
github.com/signalfx/signalfx-agent/pkg/core/config.CallConfigure(0x759c360, 0xc01a93e900, 0x76cfec0, 0xc00354fe00, 0x404d3ca, 0xc0081fc000)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/core/config/dynamic.go:121 +0x2e8
github.com/signalfx/splunk-otel-collector/internal/receiver/smartagentreceiver.(*Receiver).Start.func2()
        /Users/bjonathan/Workspace/splunk/splunk-otel-collector/internal/receiver/smartagentreceiver/receiver.go:112 +0xd3
sync.(*Once).doSlow(0xc021a8df00, 0xc002e23cb0)
        /usr/local/opt/go/libexec/src/sync/once.go:68 +0x10a
sync.(*Once).Do(0xc021a8df00, 0xc002e23cb0)
        /usr/local/opt/go/libexec/src/sync/once.go:59 +0x69
github.com/signalfx/splunk-otel-collector/internal/receiver/smartagentreceiver.(*Receiver).Start(0xc021a8dec0, 0x8427050, 0xc0001a4008, 0x8427da8, 0xa2ad9e0, 0x83dc400, 0xc00354fe00)
        /Users/bjonathan/Workspace/splunk/splunk-otel-collector/internal/receiver/smartagentreceiver/receiver.go:110 +0x7a5
github.com/signalfx/splunk-otel-collector/internal/receiver/smartagentreceiver.TestMultipleStartAndShutdownInvocations(0xc00b582780)
        /Users/bjonathan/Workspace/splunk/splunk-otel-collector/internal/receiver/smartagentreceiver/receiver_test.go:203 +0x467
testing.tRunner(0xc00b582780, 0x799c338)
        /usr/local/opt/go/libexec/src/testing/testing.go:1194 +0x203
created by testing.(*T).Run
        /usr/local/opt/go/libexec/src/testing/testing.go:1239 +0x5d8

goroutine 23161 [select]:
github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval.func1(0xc006d5e240, 0xc01d87bb80, 0x8427018, 0xc00ca9d340)
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:52 +0x174
created by github.com/signalfx/signalfx-agent/pkg/utils.RunOnInterval
        /Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/utils/time.go:46 +0x79
`

**Comment by bjsignalfx** [comment]
@rmfitzpatrick Then again, maybe not. I ran the test again and I'm not getting the error I mentioned above. I got the result below.
```
ERRO[0490]/Users/bjonathan/go/pkg/mod/github.com/signalfx/signalfx-agent@v1.0.1-0.20210218155823-9b6186460a32/pkg/monitors/cpu/cpu.go:79 github.com/signalfx/signalfx-agent/pkg/monitors/cpu.(*Monitor).generatePerCoreDatapoints() failed to calculate utilization for cpu core cpu0  error="usedDiff < 0" monitorType=cpu
2021-03-23T11:29:53.273-0400    ERROR   cpu/cpu.go:79   failed to calculate utilization for cpu core cpu0       {"component_kind": "receiver", "component_type": "smartagent", "monitorType": "cpu", "error": "usedDiff < 0"}
race: limit on 8128 simultaneously alive goroutines is exceeded, dying
exit status 66
FAIL    github.com/signalfx/splunk-otel-collector/internal/receiver/smartagentreceiver  493.321s
```

File: internal/receiver/smartagentreceiver/log.go [hunk 1]
```diff
@@ -242,3 +225,20 @@ func (l *logrusToZap) Fire(e *logrus.Entry) error {
 
 	return nil
 }
+
+func (l *logrusToZap) loggerMapValue0(src logrusKey) (*zap.Logger, bool) {
```
Comment (rmfitzpatrick) [inline internal/receiver/smartagentreceiver/log.go:229]: is this function relocation worth the change?  seems like just bumping the logrus dep and adding the test case are the helpful bits.

**Review by rmfitzpatrick — APPROVED** [review]

**Pull request merged**